tests/integration/src/compiler_test.rs

```diff
@@ -0,0 +1,379 @@
+use midenc_session::Session;
+
+enum CompilerTestSource {
+    Rust(String),
+    RustCargo {
+        cargo_project_folder_name: String,
+        artifact_name: String,
+    },
+    // Wasm(String),
+    // Ir(String),
+}
+
+/// Compile to different stages (e.g. Wasm, IR, MASM) and compare the results against expected output
+pub struct CompilerTest {
+    session: Session,
+    source: CompilerTestSource,
+    wasm_bytes: Vec<u8>,
+    hir: Option<Box<miden_hir::Program>>,
+    ir_masm: Option<Arc<miden_codegen_masm::Program>>,
+}
+
+impl CompilerTest {
+    /// Set the Rust source code to compile using a Cargo project and binary bundle name
+    pub fn rust_source_cargo(
+        cargo_project_folder: &str,
+        artifact_name: &str,
+        entrypoint: &str,
+    ) -> Self {
+        let manifest_path = format!("../rust-apps-wasm/{}/Cargo.toml", cargo_project_folder);
+        // dbg!(&pwd);
+        let temp_dir = std::env::temp_dir();
+        let target_dir = temp_dir.join(cargo_project_folder);
+        let output = Command::new("cargo")
+            .arg("build")
+            .arg("--manifest-path")
+            .arg(manifest_path)
+            .arg("--release")
+            // .arg("--bins")
+            .arg("--target=wasm32-unknown-unknown")
+            // .arg("--features=wasm-target")
+            .arg("--target-dir")
+            .arg(target_dir.clone())
+            // compile std as part of crate graph compilation
+            // https://doc.rust-lang.org/cargo/reference/unstable.html#build-std
+            .arg("-Z")
+            .arg("build-std=core,alloc")
+            .arg("-Z")
+            // abort on panic without message formatting (core::fmt uses call_indirect)
+            .arg("build-std-features=panic_immediate_abort")
+            .output()
+            .expect("Failed to execute cargo build.");
+        if !output.status.success() {
+            eprintln!("pwd: {:?}", std::env::current_dir().unwrap());
+            eprintln!("{}", String::from_utf8_lossy(&output.stderr));
+            panic!("Rust to Wasm compilation failed!");
+        }
+        let target_bin_file_path = Path::new(&target_dir)
+            .join("wasm32-unknown-unknown")
+            .join("release")
+            .join(artifact_name)
+            .with_extension("wasm");
+        // dbg!(&target_bin_file_path);
+        let mut target_bin_file = fs::File::open(target_bin_file_path).unwrap();
+        let mut wasm_bytes = vec![];
+        Read::read_to_end(&mut target_bin_file, &mut wasm_bytes).unwrap();
+        fs::remove_dir_all(target_dir).unwrap();
+
+        let session = default_session();
+
+        let entrypoint = FunctionIdent {
+            module: Ident::new(Symbol::intern("noname"), SourceSpan::default()),
+            function: Ident::new(
+                Symbol::intern(entrypoint.to_string()),
+                SourceSpan::default(),
+            ),
+        };
+        let hir_module = wasm_to_ir(&wasm_bytes, &session);
```

Comment:
I think we're going to want to use the interface provided by `midenc-compile` here (and extend it as needed). Otherwise we risk testing something other than what will be generated by the compiler in practice.

We don't need to make that change in this PR, particularly if there are things missing that make that awkward currently, but if you could make a note of the things you'd like to have from the `midenc-compile` interface, I'll make sure to get those implemented ASAP and then update these test harnesses.

Reply:
I 100% agree. I wanted to use `midenc` here initially, but I could not figure out how to get out IR and MASM without the disk and parsing text files. But, since the parsing of MASM by the Miden VM is a test of our emitter, and I'm using the disk for Rust compilation anyway, I should at least get the MASM from `midenc`. The IR part of the tests might be going away soon in the light of the upcoming `lit` tests, so I leave them as is for now if I could not figure out an easy way of getting the IR from the `midenc`.
I'd like to do all of the above in a separate PR if you don't mind since this PR is getting big, and I have other PRs stacked on top of it.

Reply:
> I'd like to do all of the above in a separate PR if you don't mind since this PR is getting big, and I have other PRs stacked on top of it.

For sure! No problem :)